tests/lang/s01_basics/s08_control_flow/t07_finally.nim

```diff
@@ -0,0 +1,786 @@
+discard """
+  targets: "c cpp"
+"""
+
+## A finally clause is executed once control-flow leaves the `finally`'s corresponding `try` and (if present) `except clause. This 
+## includes normal, interrupted (`break`, `continue`, `yield`, `return`) and exceptional (`raise`) control flow.
+
+block:
+  var values: seq[string]
+
+  proc p() =
+    block a:
+      block b:
+        try:
+          values.add("1")
+          try:
+            values.add("2")
+            return
+          finally:
+            values.add("3")
+            break a
+        finally:
+          values.add("4")
+          break b
+        values.add("5")
+      values.add("6")
+    values.add("7")
+
+  p()
+
+  doAssert values == @[
+    "1",
+    "2",
+    "3",
+    "4",
+    "6",
+    "7",
+  ]
+
+
+block:
+  var values: seq[string]
+  block a:
+    try:
+      try:
+        values.add("1")
+        raise (ref CatchableError)()
+      finally:
+        values.add("2")
+        break a
+    except:
+      values.add("3")
+    finally:
+      values.add("4")
+    values.add("5")
+
+  doAssert values == @[
+    "1",
+    "2",
+    "3",
+    "4"
+  ]
+
+block:
+  var values: seq[string]
+
+  block a:
+    block b:
+      try:
+        values.add("1")
+        break a
+      finally:
+        values.add("2")
+        block c:
+          try:
+            values.add("3")
+            break b
+          finally:
+            values.add("4")
+            break c
+          values.add("5")
+        values.add("6")
+    values.add("7")
+
+  doAssert values == @[
+    "1",
+    "2",
+    "3",
+    "4",
+    "6",
+    "7"
+  ]
+
+
+block:
+  var values: seq[string]
+  try:
+    try:
+      values.add("1")
+      raise (ref CatchableError)()
+    finally:
+      values.add("2")
+      try:
+        values.add("3")
+        raise (ref CatchableError)()
+      except:
+        values.add("4")
+  except:
+    values.add("5")
+
+  doAssert values == @[
+    "1",
+    "2",
+    "3",
+    "4",
+    "5"
+  ]
+
+
+block:
+  iterator iter(): int {.closure.} =
+    try:
+      raise (ref CatchableError)()
+    finally:
+      yield 1
+
+  var it = iter
+  # TODO: what should happen here?

```

Comment:
Per current behavior - `yield` should take priority, like any other 'jump' in finally.
Per recent discussion on how things *should* be implemented later on - `raise` should take priority.